### PR TITLE
adding spec blocks inside Move functions

### PR DESCRIPTION
**Pull request by shazqadeer**
## Motivation

This PR adds support for spec blocks inside Move functions.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/libra/libra/blob/master/CONTRIBUTING.md#pull-requests)?

Yes.

## Test Plan

Existing tests.  
Added more tests.

### CODE REVIEW

File: language/move-lang/src/parser/ast.rs [hunk 1]
```diff
@@ -211,6 +212,8 @@ pub type SpecBlockMember = Spanned<SpecBlockMember_>;
 // Specification condition kind.
 #[derive(PartialEq, Debug)]
 pub enum SpecConditionKind {
+    Assert,
```
Comment (wrwg) [inline language/move-lang/src/parser/ast.rs]: The SpecCondition was intended for pre/post conditions (that's why it is called "Condition", and invariants are a different kind of item). Should we pull this into a different item then?

Comment (shazqadeer) [reply language/move-lang/src/parser/ast.rs]: I interpreted invariants as something that are resource invariants or module invariants.  I found it cleaner to put together assert/assume together with other function-level specifications such as preconditions and postconditions, the unifying feature being that all of these are conditions that hold at specific syntactic positions in the code.  

Also note that I don't have a separate category called loop invariant.  I believe that we will be able to interpret assert/assume as asserted/assumed loop invariants just by virtue of the position where they occur.

Comment (wrwg) [reply language/move-lang/src/parser/ast.rs]: I'm fine you keep it like this, however, I think invariants, pre/post conditions, and assume/assert/loop invariants... are all different kinds of things. For instance, conditions can use `old`. Loop invariants may need other info than just the expression, we don't know yet.

File: language/move-lang/src/expansion/translate.rs [hunk 2]
```diff
@@ -24,6 +24,7 @@ struct Context {
     address: Option<Address>,
     aliases: AliasMap,
     in_spec_context: bool,
+    specs: BTreeMap<SpecId, E::SpecBlock>,
```
Comment (tnowacki) [inline language/move-lang/src/expansion/translate.rs]: Maybe label these as `exp_specs`

File: language/move-prover/spec-lang/src/translate.rs [hunk 3]
```diff
@@ -899,6 +900,8 @@ impl<'env, 'translator> ModuleTranslator<'env, 'translator> {
             let translated = et.translate_exp(exp, &BOOL_TYPE);
             et.finalize_types();
             let kind = match kind {
+                PA::SpecConditionKind::Assert => unreachable!(),
+                PA::SpecConditionKind::Assume => unreachable!(),
```
Comment (tnowacki) [inline language/move-prover/spec-lang/src/translate.rs]: How are they unreachable?

File: language/move-prover/spec-lang/src/translate.rs [hunk 4]
```diff
@@ -630,6 +630,7 @@ impl<'env, 'translator> ModuleTranslator<'env, 'translator> {
     /// Creates a SpecBlockContext from the given SpecBlockTarget
     fn get_spec_block_context(&self, target: &PA::SpecBlockTarget) -> Option<SpecBlockContext> {
         match &target.value {
+            PA::SpecBlockTarget_::Code => unreachable!(),
```
Comment (tnowacki) [inline language/move-prover/spec-lang/src/translate.rs]: How is it unreachable?

File: language/move-lang/src/expansion/translate.rs (hunk 2 above)
Comment (shazqadeer) [reply language/move-lang/src/expansion/translate.rs]: Done.

File: language/move-prover/spec-lang/src/translate.rs (hunk 3 above)
Comment (shazqadeer) [reply language/move-prover/spec-lang/src/translate.rs]: They are not.  Boogie translation of these expressions will be done in a separate PR.

File: language/move-prover/spec-lang/src/translate.rs (hunk 4 above)
Comment (shazqadeer) [reply language/move-prover/spec-lang/src/translate.rs]: They are not. Boogie translation of these expressions will be done in a separate PR.

**Comment by shazqadeer** [comment]
I have added the tests suggested by @tnowacki .

**Comment by shazqadeer** [comment]
@wrwg : I discovered that I cannot use assert as a keyword in tokenization.  So, I am now parsing assert/assume as Tok::Name.  To maintain uniformity and code cleanliness, I decided to move ensures/aborts_if also under Tok::Name.  We will revisit keywords in a later PR after discussing with the Dev Platform team.

File: language/move-lang/src/parser/syntax.rs [hunk 5]
```diff
@@ -1470,13 +1486,19 @@ fn parse_spec_block_member<'input>(tokens: &mut Lexer<'input>) -> Result<SpecBlo
...
+    let tokens_content = tokens.content();
+    let kind = if tokens_content == "assert" {
+        SpecConditionKind::Assert
+    } else if tokens_content == "assume" {
+        SpecConditionKind::Assume
+    } else if tokens_content == "aborts_if" {
         SpecConditionKind::AbortsIf
-    } else {
-        consume_token(tokens, Tok::Ensures)?;
+    } else if tokens_content == "ensures" {
```
Comment (tnowacki) [inline language/move-lang/src/parser/syntax.rs]: Switch this to a pattern match

File: language/move-lang/src/parser/syntax.rs [hunk 6]
```diff
@@ -1450,10 +1460,16 @@ fn parse_spec_block<'input>(tokens: &mut Lexer<'input>) -> Result<SpecBlock, Err
...
         Tok::NameValue => {
-            if tokens.content() == "global" {
+            let tokens_content = tokens.content();
+            if tokens_content == "assert"
+                || tokens_content == "assume"
+                || tokens_content == "aborts_if"
+                || tokens_content == "ensures"
+            {
+                parse_condition(tokens)
+            } else if tokens_content == "global" {
```
Comment (tnowacki) [inline language/move-lang/src/parser/syntax.rs]: Switch this to a pattern match

File: language/move-prover/spec-lang/src/translate.rs (hunk 3 above)
Comment (tnowacki) [reply language/move-prover/spec-lang/src/translate.rs]: I would use the macro `unimplemented!()` instead

`unreachable!()` is intended for a situation where the case/condition was checked elsewhere

File: language/move-lang/src/expansion/translate.rs [hunk 7]
```diff
@@ -877,6 +894,10 @@ fn exp_(context: &mut Context, sp!(loc, pe_): P::Exp) -> E::Exp {
             }
         }
         PE::Annotate(e, ty) => EE::Annotate(exp(context, *e), type_(context, ty)),
+        PE::Spec(spec_block) => {
+            let u = context.bind_exp_spec(spec_block);
```
Comment (tnowacki) [inline language/move-lang/src/expansion/translate.rs]: Nit, I would write `EE::Spec(context.bind_exp_spec(spec_block))` so it formats onto one line

File: language/move-lang/src/hlir/translate.rs [hunk 8]
```diff
@@ -1317,7 +1318,9 @@ fn bind_for_short_circuit(e: &T::Exp) -> bool {
     use T::UnannotatedExp_ as TE;
     match &e.exp.value {
         TE::Use(_) | TE::InferredNum(_) => panic!("ICE should have been expanded"),
-        TE::Value(_) | TE::Move { .. } | TE::Copy { .. } | TE::UnresolvedError => false,
+        TE::Value(_) | TE::Move { .. } | TE::Copy { .. } | TE::Spec(_) | TE::UnresolvedError => {
```
Comment (tnowacki) [inline language/move-lang/src/hlir/translate.rs]: Move this `TE::Spec {}` into the same spot as `Unit` below

File: language/move-lang/src/parser/syntax.rs [hunk 9]
```diff
@@ -1419,7 +1425,11 @@ fn parse_spec_block<'input>(tokens: &mut Lexer<'input>) -> Result<SpecBlock, Err
     let target = spanned(
         tokens.file_name(),
         target_start_loc,
-        tokens.previous_end_loc(),
+        if target_ == SpecBlockTarget_::Code {
+            target_start_loc
+        } else {
+            tokens.previous_end_loc()
+        },
```
Comment (tnowacki) [inline language/move-lang/src/parser/syntax.rs]: Nit I'd pattern match, but more of a personal preference here

File: language/move-lang/src/hlir/translate.rs (hunk 8 above)
Comment (shazqadeer) [reply language/move-lang/src/hlir/translate.rs]: Done.

File: language/move-lang/src/expansion/translate.rs (hunk 7 above)
Comment (shazqadeer) [reply language/move-lang/src/expansion/translate.rs]: Done.

File: language/move-lang/src/parser/syntax.rs (hunk 6 above)
Comment (shazqadeer) [reply language/move-lang/src/parser/syntax.rs]: Done.

File: language/move-lang/src/parser/syntax.rs (hunk 5 above)
Comment (shazqadeer) [reply language/move-lang/src/parser/syntax.rs]: Done.

File: language/move-lang/src/expansion/translate.rs (hunk 7 above)
Comment (shazqadeer) [reply language/move-lang/src/expansion/translate.rs]: Done.

File: language/move-prover/spec-lang/src/translate.rs (hunk 3 above)
Comment (shazqadeer) [reply language/move-prover/spec-lang/src/translate.rs]: Done.

**Pull request closed without merging**